R/tool-read-web.R

```diff
@@ -0,0 +1,220 @@
+#' @include tool-result.R
+NULL
+
+#' Tool: Read a Web Page as Markdown
+#'
+#' @examplesIf rlang::is_installed("chromote") && rlang::is_interactive()
+#' btw_tool_web_read_url("https://www.r-project.org/")
+#' btw_tool_web_read_url(
+#'   "https://posit.co/blog/easy-tool-calls-with-ellmer-and-chatlas/"
+#' )
+#'
+#' @param url The URL of the web page to read.
+#' @param ... Ignored, for future features.
+#' @param max_wait_for_page_load_s Maximum time to wait for the page to load, in
+#'   seconds. Can be set globally using the `btw.max_wait_for_page_load_s`
+#'   option.
+#'
+#' @family Tools
+#' @export
+btw_tool_web_read_url <- function(
+  url,
+  ...,
+  max_wait_for_page_load_s = getOption("btw.max_wait_for_page_load_s", 10)
+) {
+  html <- read_url_main_content(url, timeout = max_wait_for_page_load_s)
+
+  if (is.null(html) || !nzchar(html)) {
+    cli::cli_abort(
+      "Failed to read web page at {.url {url}}. Please check the URL."
+    )
+  }
+
+  md <- paste(pandoc_html_simplify(html), collapse = "\n")
+  res <- glue_('<web_page_content url="{{url}}">\n{{md}}\n</web_page_content>')
+
+  BtwWebPageResult(res)
+}
+
+BtwWebPageResult <- S7::new_class(
+  "BtwWebPageResult",
+  parent = BtwToolResult
+)
+
+has_chromote <- function() {
+  tryCatch(
+    {
+      rlang::check_installed("chromote", version = "0.5.1.9000")
+      TRUE
+    },
+    error = function(e) {
+      cli::cli_warn(
+        "The `chromote` package is required to use the web reading tool."
+      )
+      FALSE
+    }
+  )
+}
```

Comment:
When I see a function with `has_*()` prefix, I tend to assume a logical return value with no conditions raised. 

I might find usage like:

```r
if (!has_chromote()) {
   warn_no_chromote(...)
   return(whatever)
}
```

...a bit more readable.

Reply:
Done in 77df68c94294f42f3c8c5abcf20a052944a63440

Reply:
This was a good call because `btw()` should fail immediately rather than silently dropping the context anyway